google/plugin_test.go

```diff
@@ -1,82 +1,82 @@
-			t.Errorf("Parsing non-Google upstream should not have errors %v", err)
-		}
-	}
-}
+var _ = Describe("Plugin", func() {
+	Describe("Processing upstream", func() {
+		Context("With non-Google upstream", func() {
+			It("should not error and return nothing", func() {
+				upstreams := []*v1.Upstream{&v1.Upstream{}, &v1.Upstream{Type: "some-upstream"}}
+				p := Plugin{}
+				for _, u := range upstreams {
+					err := p.ProcessUpstream(nil, u, nil)
+					Expect(err).NotTo(HaveOccurred())
+				}
+			})
+		})
 
-func TestProcessingGoogleUpstream(t *testing.T) {
-	upstream := &v1.Upstream{
-		Type: UpstreamTypeGoogle,
-		Spec: upstreamSpec("us-east1", "project-x"),
-	}
-	out := &envoyapi.Cluster{}
-	p := Plugin{}
-	err := p.ProcessUpstream(nil, upstream, out)
-	if err != nil {
-		t.Errorf("parsing valid upstream shouldn't have errors %v", err)
-	}
-	if !p.isNeeded {
-		t.Errorf("google plugin should be required after valid google upstream")
-	}
-	if len(out.Hosts) < 1 {
-		t.Errorf("was expecting hosts")
-	}
-	address, ok := out.Hosts[0].Address.(*envoycore.Address_SocketAddress)
-	if !ok {
-		t.Errorf("was expecting address")
-	}
-	if !strings.Contains(address.SocketAddress.Address, "us-east1") {
-		t.Error("was expecting region in hostname")
-	}
-	if !strings.Contains(address.SocketAddress.Address, "project-x") {
-		t.Error("was expecting project in hostname")
-	}
-}
+		Context("with valid upstream spec", func() {
+			var (
+				err error
+				out *envoyapi.Cluster
+			)
 
-func TestProcessingFunctionSpecForNonGoogle(t *testing.T) {
-	p := Plugin{}
-	nonGoogle := &plugin.FunctionPluginParams{}
-	out, err := p.ParseFunctionSpec(nonGoogle, funcSpec("https://host.io/func"))
-	if err != nil {
-		t.Errorf("non google params should not result in error")
-	}
-	if out != nil {
-		t.Errorf("non google params should return nil")
-	}
-}
+			BeforeEach(func() {
+				upstream := &v1.Upstream{
+					Type: UpstreamTypeGoogle,
+					Spec: upstreamSpec("us-east1", "project-x"),
+				}
+				out = &envoyapi.Cluster{}
+				p := Plugin{}
+				err = p.ProcessUpstream(nil, upstream, out)
+			})
 
-func TestProcessingFunctionSpecForGoogle(t *testing.T) {
-	p := Plugin{}
-	nonGoogle := &plugin.FunctionPluginParams{UpstreamType: UpstreamTypeGoogle}
-	out, err := p.ParseFunctionSpec(nonGoogle, funcSpec("https://host.io/func"))
-	if err != nil {
-		t.Errorf("unexpected error %v", err)
-	}
-	if host := get(out, functionHost); host != "host.io" {
-		t.Error("host didn't match", host)
-	}
-	if path := get(out, functionPath); path != "/func" {
-		t.Error("path didn't match", path)
-	}
-}
+			It("should not error", func() {
+				Expect(err).NotTo(HaveOccurred())
+			})
+
+			It("should have a host in output", func() {
+				Expect(out.Hosts).Should(HaveLen(1))
+				Expect(out.Hosts[0].GetSocketAddress()).NotTo(BeNil())
+			})
+
+			It("should have region and project in the output host", func() {
+				Expect(out.Hosts[0].GetSocketAddress()).To(ContainSubstring("us-east1"))
```

Comment:
consider using
out.Hosts[0].GetSocketAddress().Address
out.Hosts[0].GetSocketAddress().PortSpecifier.(*envoycore.SocketAddress_PortValue).PortValue
to be more explicit